### ISSUE TITLE
Enhancement: [no-non-null-assertion] Optionally exempt contexts where null would immediately throw at runtime

### ISSUE THREAD

**Issue by andersk** [typescript-eslint/typescript-eslint#9168]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-non-null-assertion

### Description

`no-non-null-assertion` is a good rule because `obj!` has no runtime check and blows a dangerous hole in type safety: if a bug causes `obj` to be `null` or `undefined` at runtime, the ill-typed value could propagate far into the code and cause an problem elsewhere that’s difficult to track down to its original cause.

However, for some *specific contexts* such as `obj!.prop`, there is no such propagation of ill-typed values, because `null` or `undefined` in that context would immediately throw a `TypeError` at runtime. In that sense, `obj!.prop` is just as safe as the well-typed `assert(obj); obj.prop` or `nullthrows(obj).prop` that would typically replace it. But `obj!.prop` is simpler, slightly more efficient, and has the bonus advantage of allowing the JS runtime to produce a better error message by default (e.g. Firefox gives `TypeError: obj is undefined` with the actual name of the variable `obj`).

I propose a new option `allowWhenNullThrows` to `no-non-null-assertion` that allows the non-null assertion in certain contexts such as `obj!.prop` where `null` or `undefined` can’t break type safety because it would immediately throw at runtime. I’ve enumerated a number of such contexts below.

### Fail

```TypeScript
Everything currently rejected by `@typescript-eslint/no-non-null-assertion`, with the few optional exemptions below.
```

### Pass

```TypeScript
With `"@typescript-eslint/no-non-null-assertion": ["error", {allowWhenNullThrows: true}]`:


declare let o: { p: number } | null | undefined,
  a: number[] | null | undefined,
  f: ((u: unknown) => void) | null | undefined,
  c: (new () => void) | null | undefined;

o!.p; // property access
"p" in o!; // property test
let { p } = o!; // object destructuring declaration
({ p } = o!); // object destructuring assignment
with (o!) { } // with (though it’s forbidden in TS anyway)
a![0]; // indexing
[...a!]; // array spread (but not object spread)
Array(...a!); // argument spread
let [n] = a!; // array destructuring declaration
[n] = a!; // array destructuring assignment
(function* () { yield* a!; }); // generator delegation
f!(0); // function call
f!`t`; // tagged template literal
new c!(); // constructor
({}) instanceof c!; // constructor test
```

**Comment by bradzacher** [comment]
The point isn't to ban it so that you use a throws assertion - though that is one remedy to the rule.

For example one could change `x!.y` to `assert(x).y`. But another option - a better option in many cases - is to handle that case specifically eg by returning early instead. Another option again is to use optional chaining.

---

Ultimately the reason you want to avoid the non-null assertion is because you want to completely avoid these unexpected runtime crashes. 

For a lot of people the whole idea behind TS is that if your code typechecks then it is unlikely to unexpectedly crash at runtime because **unexpected** runtime crashes are going to impact your users.
But a non-null assertion breaks that expectation because it's an inherently unsafe operator.

---

Personally I don't think that this request aligns with the intent or goals of the rule, nor do I think that this would be broadly applicable to the community.

I'd like to hear from the rest of the team though @typescript-eslint/triage-team

**Comment by kirkwaiblinger** [comment]
Can you provide an example of what _would_ still be checked by the rule?

**Comment by kirkwaiblinger** [comment]
I guess 
```ts
let y: string | undefined = undefined;
let x: string = y!;
```
But what else that's nontrivial survives?

**Comment by kirkwaiblinger** [comment]
(The reason I ask is not to be difficult, but I'm having a hard time understanding what the rule _does_ do in this scenario. I can understand if you don't want to use the rule at all, but I'm having a harder time understanding what value the rule still provides after omitting the cases you enumerate, so I would love if you could spell a few examples out 🙏)

**Comment by bradzacher** [comment]
That's actually a really good point. Likely almost all the cases where there's not a ***direct*** runtime error, there is instead going to be an ***indirect*** runtime error.

Eg

```ts
// would be reported by your option
const x = thing!;
x.y;

// would not be reported by your option
thing!.y;
```

So really you'd need to spell out exactly what cases are expected to report. Our hypothesis is that those cases will have indirect runtime errors. And at that point it would follow that really you just want to allow non-null assertions because you're okay with the unexpected runtime errors they introduce.

**Comment by andersk** [comment]
The proposed option wouldn’t exempt anything that isn’t an **immediate** runtime error. The contexts I listed under “Pass” are the only contexts I’m proposing to exempt with this option, unless someone thinks of other contexts with the same property.

These violations would still be reported (this is not an exhaustive list):

```ts
const x = thing!; // still reported
x.y;

func(arg!); // still reported
arr[n!]; // still reported
n! + 1; // still reported
```

**Comment by kirkwaiblinger** [comment]
I see, so is it fair to say that the point is largely that these cause _likely silent_ logic errors, but not loud exceptions, which makes them _more_ nefarious than the loud exceptions? (and if an exception _does_ result, it is less co-located with the type-unsafe syntax than the provably immediately-throwing cases)

**Comment by bradzacher** [comment]
Yeah so that confirms what we're saying. Whilst they don't directly cause runtime errors they no doubt will indirectly cause them - maybe just a line or two later.

Eg `x[y!].z` - the non-null assertion itself won't cause a runtime error as a null/undefined would be coerced to a string and accessed as a key. But likely the key would not exist - so then the following `.z` would immediately fail at runtime.

I just don't see a reason why you would want to allow a specific subset of runtime crashes into your code and ban another subset of them.

I don't see that there is broad applicability to this option either - people turn this on because they want to enforce safer code and I don't think allowing a whole lot of unsafety aligns with that ideal.

For these reasons, I'm going to reject this request. Thanks for filing and discussing!

**Comment by andersk** [comment]
I only want to exempt things that are **guaranteed** to **immediately** fail at runtime.

`x[y!].z` is not guaranteed to immediately fail at runtime, so it would still be reported. My proposal is based on hard guarantees, not some heuristic of “likely”.

The goal here isn’t to prevent all runtime crashes—if we wanted that, we would need to ban (at least) `throw`, `assert`, and `nullthrows` by exactly the same logic. The goal is to prevent **non-local** runtime crashes, caused by propagation of a value whose runtime type doesn’t match its static type. Errors are unavoidable, and execution will abort on errors; I just want my errors to point directly at the problem, and I want my types to match along all paths where execution might proceed.

**Comment by kirkwaiblinger** [comment]
It's a thoroughly thought out and knowledgeable proposal, to be sure! (And, I do follow what you're saying regarding the point that it's not a "likeliness" heuristic). And, I am convinced by this proposal that some cases of unsound nonnull assertions are more nefarious than others.

That being said, I am -1 as well... 

I agree fully with each of the other team members' comments. I would add also the following thoughts for my position:

1. it's not intuitive for anyone other than the person who configures the eslintrc for their team's codebase (people like you and me 😄, but not "broadly applicable")
2. Related to this, I think it creates lots of potential for heinous workarounds/user confusion around what is and isn't sound. For example, would you consider this a valid resolution to the bug the concern that the linter is flagging?
   ```ts
   // Failed the linter! This must be Bad Code
   const definitelyString: string = maybeString!;
   
   // Passed the linter! This must be Good Code
   const definitelyString: string = maybeString!.valueOf();
   ```
   I'd argue it satisfies both the spirit and the letter of the proposal. But is very 😬 in my opinion. Why wouldn't `nullThrows(maybeString)` be hugely preferable here?
3. Yes, errors occur in code. But, I cannot be convinced that postfix `!`, and only when used in a specific subset of syntactic positions, is a reasonable way of throwing one. The _whole point_ of the nonnull assertion operator is to tell the compiler "I don't care if downstream code has errors if this is nullish; I _know_ this is safe enough to not even bother checking", not "please throw in a debuggable way if this is nullish". That's the use case for a `nullThrows`-with-required-custom-message helper.

At the end of the day, my honest feeling is that the problem this option would solve (`nullThrows(x)` is not ergonomic and requires a second argument to give good error messages) is not significant enough, and is not fully solved enough by the proposed option (due to varying JS runtimes and bundler/minification-dependent behavior), to justify the confusing drawbacks that it introduces, for the purposes of a general-purpose lint rule.
